packages/indexer-common/src/indexer-management/allocations.ts

```diff
@@ -0,0 +1,1134 @@
+    logger: Logger,
+    allocationID: string,
+    poi: string | undefined,
+    amount: BigNumber,
+    force: boolean,
+  ): Promise<PopulatedTransaction> {
+    const params = await this.prepareReallocateParams(
+      logger,
+      allocationID,
+      poi,
+      amount,
+      force,
+    )
+    return await this.contracts.staking.populateTransaction.closeAndAllocate(
+      params.closingAllocationID,
+      params.poi,
+      params.indexer,
+      params.subgraphDeploymentID,
+      params.tokens,
+      params.newAllocationID,
+      params.metadata,
+      params.proof,
+    )
+  }
+
+  async reallocate(
+    allocationID: string,
+    poi: string | undefined,
+    amount: BigNumber,
+    force: boolean,
+  ): Promise<ReallocateAllocationResult> {
+    try {
+      const params = await this.prepareReallocateParams(
+        this.logger,
+        allocationID,
+        poi,
+        amount,
+        force,
+      )
+
+      this.logger.info(`Sending closeAndAllocate transaction`, {
+        indexer: params.indexer,
+        oldAllocation: params.closingAllocationID,
+        newAllocation: params.newAllocationID,
+        newAllocationAmount: formatGRT(params.tokens),
+        deployment: params.subgraphDeploymentID,
+        poi: params.poi,
+        proof: params.proof,
+      })
+
+      const receipt = await this.transactionManager.executeTransaction(
+        async () =>
+          this.contracts.staking.estimateGas.closeAndAllocate(
+            params.closingAllocationID,
+            params.poi,
+            params.indexer,
+            params.subgraphDeploymentID,
+            params.tokens,
+            params.newAllocationID,
+            params.metadata,
+            params.proof,
+          ),
+        async (gasLimit) =>
+          this.contracts.staking.closeAndAllocate(
+            params.closingAllocationID,
+            params.poi,
+            params.indexer,
+            params.subgraphDeploymentID,
+            params.tokens,
+            params.newAllocationID,
+            params.metadata,
+            params.proof,
+            { gasLimit },
+          ),
+        this.logger.child({ function: 'staking.closeAndAllocate' }),
+      )
+
+      return await this.confirmReallocate(0, allocationID, receipt)
+    } catch (error) {
+      this.logger.error(error.toString())
+      throw error
+    }
+  }
+
+  async matchingRuleExists(
+    logger: Logger,
+    subgraphDeploymentID: SubgraphDeploymentID,
+  ): Promise<boolean> {
+    const indexingRules = await fetchIndexingRules(this.models, true)
+    const subgraphDeployment = await this.networkMonitor.subgraphDeployment(
+      subgraphDeploymentID.ipfsHash,
+    )
+    if (!subgraphDeployment) {
+      throw Error(
+        `SHOULD BE UNREACHABLE: No matching subgraphDeployment (${subgraphDeploymentID.ipfsHash}) found on the network`,
+      )
+    }
+    return isDeploymentWorthAllocatingTowards(logger, subgraphDeployment, indexingRules)
```

Comment:
I'm not sure I see much of a difference in readability either way. What do you think?

One nice thing about the way it currently is would be that it can eventually replace all of https://github.com/graphprotocol/indexer/blob/main/packages/indexer-common/src/network.ts#L335-L417.

Reply:
The naming `matchingRuleExists` seem to suggest that there would be a stricter check on indexing rules when it currently just check if the deployment id is found on the network while it grab all the indexing rules. 

To be a bit more specific, I thought it would make sense to 
- move the rule finding at [network.ts#L335-L344](https://github.com/graphprotocol/indexer/blob/main/packages/indexer-common/src/network.ts#L335-L344) to here
- we remove the check on 346 for altruistic allocations ([GGP 0012] (https://snapshot.org/#/council.graphprotocol.eth/proposal/0xd64e523713a05682f6cef9bfaf7fb340dde3396aeb1fe16d6ec5c1e1eaa7dd1e))
- we include the check on `L356` and filter for `IndexingDecisionBasis.ALWAYS` before conditionally calling `isDeploymentWorthAllocatingTowards`.

But it is also fine to leave it and make a separate PR later as you are saying that this function should eventually replace all of that section :)